### PR TITLE
IPC-FBP Customization

### PR DESCRIPTION
**Pull request by KasparoVy**
Inspired by Polaris FBPs, added in **Zuhayr**'s huge IPC overhaul a while back. The goal of this was to enable enhanced customization without detracting or changing the current mode of IPC customization in any way. 
This means that IPC customization as it is in the server right now without this PR is just the same in this PR. Only when you choose the other body parts will you unlock the other customization options.

Huge thanks to all contributors, namely **Tigercat2000** for helping me un-fuck pretty much the entire PR, **Spacemanspark** for testing and the antennae suggestion, **TheDZD**, **Tastyfish** and **Crazylemon64** for their suggestions, help and proofreading as well. Anyone else who assisted and I forgot to name, I apologize

**Adds:**
- The ability to change your head, torso and groin as an IPC. Non-IPCs cannot access the parts.
- Head, torso and groin sprites for all mechanical limb/bodypart brands but Morpheus (since those already exist) from Polaris.
- Faux-eye optics for non-Morpheus heads.
- The ability to change optic (eye) colour if you've got a non-Morpheus head.
- The ability to choose human hair styles (wigs) and facial hair styles (postiche) with non-Morpheus heads. Non-Morpheus heads cannot choose screens. Conversely, Morpheus heads cannot choose wigs or facial hair.
- Two human hair styles from Polaris.
- The ability for IPCs to wear undergarments.
- Antennae for IPCs.

**Tweaks:**
- IPC monitor adjustment verb to adjust optics if the head is non-Morpheus.
- Appearance changing proc slightly tweaked in order to ensure IPCs with non-Morpheus heads can't get screens from a mirror, and so IPCs with Morpheus (default) heads can't get hair from mirrors.

:cl:
rscadd: The ability to change your head, torso and groin as an IPC in the character creator. Non-IPCs cannot access the parts.
rscadd: Head, torso and groin sprites for all mechanical limb/bodypart brands but Morpheus and the unbranded ones (since those already exist) from Polaris.
rscadd: Faux-eye optics for non-Morpheus heads.
rscadd: The ability to change optic (eye) colour if you've got a non-Morpheus head.
rscadd: The ability to choose human hair styles (wigs) and facial hair styles (postiche) with non-Morpheus heads. Non-Morpheus heads cannot choose screens. Conversely, Morpheus heads cannot choose wigs or facial hair.
rscadd: Two hair styles from Polaris.
rscadd: The ability for IPCs to wear undergarments (shirts and trousers).
rscadd: Antennae (colourable) for IPCs.
tweak: IPC monitor adjustment verb will now adjust optic colour if the head is non-Morpheus.
/:cl:

### CODE REVIEW

**Comment by KasparoVy** [comment]
_**Example of additional customization features in creation UI**_
ipccustomizationui1

_**Proof that you can still have the same ol' regular monitor-head IPCs**_
ipccustomizationui2

_**IPCs with Morpheus heads can only use screen hair-styles. (Character creation)**_
ipccustomizationui3

_**Conversely, IPCs with any of the non-Morpheus heads cannot use screen styles. But instead get the ability to wear fake hair (head and facial)**_
ipccustomizationui4
ipccustomizationui5

_**Example of an ID for an IPC with customized body/head (Didn't have to change anything in datacore.dm at all to get this to work)**_
ipccustomizationid1

_**IN-GAME EXAMPLES**_
ipccustomizationig1

_**Legacy customization still works the exact same**_
ipccustomizationig2

_**Using the change monitor/optics verb when you've got a non-Morpheus head (You can only change your optic (eye) colour**_
ipccustomizationig3

_**Same verb but with a Morpheus head (Works the same as before, only IPC screens listed here)**_
ipccustomizationig5

_**Using the mirror**_
ipccustomizationig4

ipccustomizationig6

**Comment by KasparoVy** [comment]
@Spacemanspark Thank you very very much for the suggestion.

See if you can pick out all of the part manufacturers! The first two on the left are a mix, however.

_**Antennae and some body previews**_
ipccustomizationig8

The only one I didn't include a full preview of was the Zeng-Hu prosthetics (because I forgot, woops). The Zeng-hu head is visible on one of these models, however.

And that's without even messing with the body colour, hahaha.

**Comment by Spacemanspark** [comment]
Wheeeee

**Comment by DaveTheHeadcrab** [comment]
Much distaste as I have for some of the more clique-ey IPC players, I think this is cool as hell.
:+1:

**Comment by NTSAM** [comment]
Still :+1: , cool as fuck.

**Comment by FlattestGuitar** [comment]
I still dislike the human-looking heads, but the rest is amazing. Good job on porting! 

:+1: / :+1:

**Comment by TheDZD** [comment]
Merge conflicts.

**Comment by Dumbdumn5** [comment]
This makes me want to sprite some unique alt heads for each of the individual corporations
Will get to work on it ASAP, will have to add in some cool lights too. The bodies are too good to not have a cool face to go with it!

**Comment by KasparoVy** [comment]
At work, will deal with this later.

**Comment by tigercat2000** [comment]
@KasparoVy quick tip: you don't need to close and reopen a PR to fix merge conflicts
Once you get a branch and fix the merge conflicts; Switch back to the PR branch, and reset it to the branch you fixed (i assume that's why you make new PR's, because of branches)
To reset a branch, you run `git reset --hard branch_name_here`

Then, you can overwrite the stuff in the current pushed branch, by running `git push origin -f pr_branch_name_here` (the -f signals "force", as in, overwrite any commits, allow non-fast-forward tree changes)

**Comment by Dumbdumn5** [comment]
Got a mock-up bishop alt head front done, ere'.

[http://imgur.com/R2ku4hQ](url)

**Comment by Dinarzad** [comment]
That bottom right one in the last picture set is awesome as hell
It reminds me a LOT of the [Adjutants](http://vignette1.wikia.nocookie.net/starcraft/images/a/a4/Adjutant_SC2_Head2.jpg/revision/latest?cb=20100906202000) from the Starcraft series, and I will totally make one.

**Comment by fludd12** [comment]
And in that moment,
we were all warframes
*beep

**Comment by KasparoVy** [comment]
Alright, just got home. Going to sort out the conflicts

Thanks so much @tigercat2000 for explaining that to me, you were right in your assumption
Time to sort some conflicts

**Comment by KasparoVy** [comment]
Alright, the conflicts are resolved but now the PR has a commit with a kek name because I'm an idiot
Lmao. At least I know wat it do now.

**Comment by Spacemanspark** [comment]
>Merge when?

**Comment by NTSAM** [comment]
>>Merge when?

**Comment by ScratKnapp** [comment]
> Merge when?

**Comment by tigercat2000** [comment]
@Fox-McCloud @TheDZD conflicts on here fixed, it's probably fine to merge

**Comment by Spacemanspark** [comment]
> > > >[insert meme here]

**Comment by NTSAM** [comment]
>>>>>Merge when?>wen?

**Comment by Spacemanspark** [comment]
I know I'm being a bit impatient, but can this get merged/ reviewed if it hasn't been?  
It's waited around ten days.

**Comment by NTSAM** [comment]
>I know I'm being a bit impatient, but can this get merged/ reviewed if it hasn't been?
>It's waited around ten days.

File: code/modules/admin/admin_verbs.dm [hunk 1]
```diff
@@ -809,7 +809,17 @@ var/list/admin_verbs_proccall = list (
 		return
 
 	if(!istype(H))
-		return
+		if(istype(H, /mob/living/carbon/brain))
```
Comment (tigercat2000) [inline code/modules/admin/admin_verbs.dm]: Why would anyone ever open the CMA menu on a brain?

File: code/modules/admin/admin_verbs.dm [hunk 2]
```diff
@@ -825,7 +835,17 @@ var/list/admin_verbs_proccall = list (
...
```
Comment (tigercat2000) [inline code/modules/admin/admin_verbs.dm]: same here, wat

File: code/modules/admin/admin_verbs.dm (hunk 1 above)
Comment (Spacemanspark) [inline code/modules/admin/admin_verbs.dm]: It's a crazy world/ space station.

File: code/modules/mob/living/carbon/human/human.dm [hunk 3]
```diff
@@ -1619,26 +1619,36 @@
...
+	if(species.flags & ALL_RPARTS) //If they can have a fully cybernetic body...
+		if(client.prefs.rlimb_data["head"]) //If head is present here, that means it's not the default Morpheus. Thus, no screen to adjust. Instead, let them change the colour of their optics!
+			var/optic_colour = input(src, "Select optic colour", rgb(r_markings, g_markings, b_markings)) as color|null
+			if(optic_colour)
+				r_markings = hex2num(copytext(optic_colour, 2, 4))
+				g_markings = hex2num(copytext(optic_colour, 4, 6))
+				b_markings = hex2num(copytext(optic_colour, 6, 8))
+
+			update_markings()
+		else if(!(client.prefs.rlimb_data["head"]))//Means that the character has the default Morpheus head, which has a screen. Time to customize.
```
Comment (tigercat2000) [inline code/modules/mob/living/carbon/human/human.dm]: excess brackets
`else if(!client.prefs.rlimb_data["head"])`

File: code/modules/mob/living/carbon/human/appearance.dm [hunk 4]
```diff
@@ -192,9 +219,23 @@
...
+					continue
+			else
+				if(!user.client.prefs.rlimb_data["head"]) //If the facial hair style is not native to the user's species, and they're using the default head, don't let them access it.
+					continue
+				else
+					if(("Human" in S.species_allowed)) //If the user has a robotic head and the facial hair style can fit humans, let them use it as a postiche for their humanoid robot head.
+						valid_facial_hairstyles += facialhairstyle
+					continue
+		else //If the user is not a species who can have robotic heads, use the default handling.
+			if(!(species.name in S.species_allowed))
```
Comment (tigercat2000) [inline code/modules/mob/living/carbon/human/appearance.dm]: Okay. I think it's worth explaining.

To the best of my knowledge, the `in` operator only behaves really oddly when it is in a multi-part if statement.
When it's used alone,
`if(species.name in S.species_allowed)` or `if(!species.name in S.species_allowed`, it generally works fine-
From what I know, at least.

File: code/modules/admin/admin_verbs.dm (hunk 1 above)
Comment (KasparoVy) [inline code/modules/admin/admin_verbs.dm]: If you look further down, it won't work for just any brain. It checks if it's an MMI'd brain, then if it's got a body. If it doesn't, it won't do jack.

If I didn't have this, it flat-out wouldn't work for IPCs, morpheus heads or no. It was so wierd.

File: code/modules/admin/admin_verbs.dm (hunk 2 above)
Comment (KasparoVy) [inline code/modules/admin/admin_verbs.dm]: Same as above. Only added out of necessity, it wouldn't work any other way and I tried to make it such that it wouldn't permit any un-mmi'd brains without IPC bodies from doing this.

File: code/modules/mob/living/carbon/human/appearance.dm (hunk 4 above)
Comment (KasparoVy) [inline code/modules/mob/living/carbon/human/appearance.dm]: Thanks, I figured from what you said before I'd be better off taking the 'rather safer than sorry(er)' approach, but I'll strip the extra brackets.

File: code/modules/admin/admin_verbs.dm (hunk 1 above)
Comment (tigercat2000) [inline code/modules/admin/admin_verbs.dm]: that shouldn't be possible in any way, what the fuck byond

**Comment by KasparoVy** [comment]
Everything still works just fine

File: code/modules/client/preferences.dm [hunk 5]
```diff
@@ -1103,7 +1109,7 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 								continue
 							if(gender == FEMALE && S.gender == MALE)
 								continue
-							if( !(species in S.species_allowed))
+							if(!(species in S.species_allowed))
```
Comment (KasparoVy) [inline code/modules/client/preferences.dm]: I need the extra brackets in places like this for example because if I use !species in S.species_allowed it borks hardcore in the sense that instead of only showing intended hairstyles, it'll show pretty much all hairstyles except for (I think) the first species in the list that doesn't match the species that's being searched for or something.

tl;dr: if I have it like if(!species in S.species_allowed) it breaks, if I leave it like this it's good

File: code/modules/client/preferences.dm [hunk 6]
```diff
@@ -350,7 +356,7 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 				dat += "[TextPreview(flavor_text)]...<br>"
 			dat += "<br>"
 
-			if((species in list("Unathi", "Vulpkanin", "Tajaran"))) //Species that have head accessories.
+			if(species in list("Unathi", "Vulpkanin", "Tajaran", "Machine")) //Species that have head accessories.
```
Comment (KasparoVy) [inline code/modules/client/preferences.dm]: removing this set of brackets caused no issues whatsoever, they were just extra and I removed them like I was told to do with extra brackets

File: code/modules/client/preferences.dm (hunk 5 above)
Comment (tigercat2000) [inline code/modules/client/preferences.dm]: how odd

**Comment by KasparoVy** [comment]
I guess I'll just keep checking back to make sure this isn't conflicting with anything.

**Comment by Spacemanspark** [comment]
Well, Fox did say the PR itself was fine, now that I think about it, and the sprites look good.

**Comment by Fox-McCloud** [comment]
@TheDZD

File: code/modules/client/preferences.dm (hunk 5 above)
Comment (Crazylemon64) [inline code/modules/client/preferences.dm]: Yeah, I think it'd group like ((!species) in S.species_allowed), which is very definitely going to smash things like crazy. The more brackets, the better, when it comes to things that will hilariously explode on a bad regrouping.

Comment (KasparoVy) [inline code/modules/client/preferences.dm]: That's what I figured but I wasn't entirely sure until I tested it, I'm glad someone else is able to confirm this haha

**Comment by TheDZD** [comment]
Fingers crossed that we don't unleash some Eldritch horror from customization code.

**Comment by NTSAM** [comment]
:clap:

**Comment by Crazylemon64** [comment]
N'yarla'gozoth yggl bxx'zl gzzlthoth. N'gliboth boodla omniqual azerroth. 'ia! 'ia!

**Comment by NTSAM** [comment]
Help, I think Crazy is either having a seizure or is a follower of the Cult of Cthulu!

**Comment by Spacemanspark** [comment]
Yay.

File: code/modules/client/preferences.dm [hunk 7]
```diff
@@ -1392,37 +1453,50 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
...
-							if(second_limb)
-								organ_data[second_limb] = "amputated"
-								rlimb_data[second_limb] = null
+							if(!no_amputate)
+								organ_data[limb] = "amputated"
+								rlimb_data[limb] = null
+								if(second_limb)
+									organ_data[second_limb] = "amputated"
+									rlimb_data[second_limb] = null
 						if("Prothesis")
```
Comment (KasparoVy) [inline code/modules/client/preferences.dm]: Fixed this typo in #3943, supposed to be Prosthesis

prothesis proboscis PNODSFPOGNGINS[D
